### PR TITLE
initial code commit

### PR DESCRIPTION
**Pull request by HarshalRane23**
Hi Team,

As suggested by Morgante Pell I have committed my code to this [repo](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/master/examples) in the example section. so it can help other in implementing terraform. This is first time I am adding code to public repo, let me know if ant changes needed
Issue [link](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/587)

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_whitespace
 Failed whitespace check. More details below. 
 ```bash 
Checking for trailing whitespace
./examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_private_cluster/main.tf:42:  
./examples/windows_nodepool_with_private_cluster/modules/README.md:1:# Terraform GCP Resource Module 
./examples/windows_nodepool_with_private_cluster/gke/outputs.tf:2:  value     = module.gke_private_cluster.name  
./examples/windows_nodepool_with_private_cluster/gke/outputs.tf:6:  value     = module.gke_private_cluster.region  
./examples/windows_nodepool_with_private_cluster/gke/main.tf:10:  ip_range_services          = "service-range"  # e.g : 10.2.4.0/24 
./examples/windows_nodepool_with_private_cluster/gke/main.tf:27:# Linux Nodepool 
./examples/windows_nodepool_with_private_cluster/gke/main.tf:54:# Windows Nodepool 
./examples/windows_nodepool_with_private_cluster/README.md:1:## 
./examples/windows_nodepool_with_private_cluster/README.md:16:| region | The region to host the cluster in | string | n/a | yes | 
./examples/windows_nodepool_with_private_cluster/README.md:17:| zones | The zone to host the cluster in | string | n/a | yes | 
Error: Trailing whitespace found in the lines above.
Checking for missing newline at end of file
Error: No newline at end of file ./examples/windows_nodepool_with_private_cluster/modules/README.md
Error: No newline at end of file ./examples/windows_nodepool_with_private_cluster/gke/backend.tf
Error: No newline at end of file ./examples/windows_nodepool_with_private_cluster/gke/variables.tf
 ``` 
- ⚠️check_headers
 All files need a license header. Please make sure all your files include the appropriate header. A helper tool available [here](https://github.com/google/addlicense). 
 ```bash 
Checking file headers
6 files have incorrect boilerplate headers:
examples/windows_nodepool_with_private_cluster/gke/backend.tf
examples/windows_nodepool_with_private_cluster/gke/main.tf
examples/windows_nodepool_with_private_cluster/gke/outputs.tf
examples/windows_nodepool_with_private_cluster/gke/providers.tf
examples/windows_nodepool_with_private_cluster/gke/variables.tf
examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_node_pool/variables.tf
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_private_cluster/main.tf
--- old/examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_private_cluster/main.tf
+++ new/examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_private_cluster/main.tf
@@ -38,8 +38,8 @@
   // for regional cluster - use var.zones if provided, use available otherwise, for zonal cluster use var.zones with first element extracted
   // node_locations = var.regional ? coalescelist(compact(var.zones), slice(data.google_compute_zones.available.names, 0, 3)) : slice(var.zones, 0, length(var.zones))
   node_locations = var.regional ? coalescelist(compact(var.zones), slice(data.google_compute_zones.available.names, 0, 3)) : null
-  zones = var.regional ? local.node_locations : var.zones
-  
+  zones          = var.regional ? local.node_locations : var.zones
+
   // Kubernetes version
   master_version_regional = var.kubernetes_version != "latest" ? var.kubernetes_version : data.google_container_engine_versions.region.latest_master_version
   master_version_zonal    = var.kubernetes_version != "latest" ? var.kubernetes_version : data.google_container_engine_versions.zone.latest_master_version
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
examples/windows_nodepool_with_private_cluster/gke/outputs.tf
--- old/examples/windows_nodepool_with_private_cluster/gke/outputs.tf
+++ new/examples/windows_nodepool_with_private_cluster/gke/outputs.tf
@@ -1,9 +1,9 @@
 output "kubernetes_cluster_name" {
-  value     = module.gke_private_cluster.name  
+  value = module.gke_private_cluster.name
 }
 
 output "kubernetes_cluster_region" {
-  value     = module.gke_private_cluster.region  
+  value = module.gke_private_cluster.region
 }
 
 
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
examples/windows_nodepool_with_private_cluster/gke/backend.tf
--- old/examples/windows_nodepool_with_private_cluster/gke/backend.tf
+++ new/examples/windows_nodepool_with_private_cluster/gke/backend.tf
@@ -1,6 +1,6 @@
 #terraform {
- # backend "gcs" {
-  #  bucket = "bucket-name"
-   # prefix = "folder-prefix"
-  #}
+# backend "gcs" {
+#  bucket = "bucket-name"
+# prefix = "folder-prefix"
+#}
 #}
\ No newline at end of file
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
examples/windows_nodepool_with_private_cluster/gke/main.tf
--- old/examples/windows_nodepool_with_private_cluster/gke/main.tf
+++ new/examples/windows_nodepool_with_private_cluster/gke/main.tf
@@ -7,7 +7,7 @@
   network                    = var.network
   subnetwork                 = var.subnetwork
   ip_range_pods              = "pod-range"     # Note : pod-range is name of additional range in subnet. e.g : 10.2.0.0/16 or 10.2.4.0/21 (select wider range for pods as autoscalling will have large ip Requirements)
-  ip_range_services          = "service-range"  # e.g : 10.2.4.0/24 
+  ip_range_services          = "service-range" # e.g : 10.2.4.0/24 
   regional                   = var.regional
   zones                      = var.zones
   http_load_balancing        = true
@@ -52,28 +52,28 @@
 }
 
 # Windows Nodepool 
- module "gke_node_pool_02" {
-   source                         = "../modules/google_kubernetes_engine/gke_node_pool/"
-   project_id                     = var.project
-   gke_cluster_name               = module.gke_private_cluster.name
-   node_pool_name                 = "windows-node-pool"
-   region                         = var.region
-   regional                       = var.regional
-   zones                          = module.gke_private_cluster.zones
-   gke_cluster_min_master_version = var.gke_cluster_min_master_version
-   image_type                     = "WINDOWS_LTSC"
-   machine_type                   = var.node_pool_02["machine_type"]
-   preemptible                    = false
-   auto_upgrade                   = false
-   auto_repair                    = true
-   max_pods_per_node              = "100"
-   node_count                     = "1"
-   local_ssd_count                = "0"
-   enable_autoscaling             = true
-   min_node_count                 = var.node_pool_02["min_node_count"]
-   max_node_count                 = var.node_pool_02["max_node_count"]
-   labels                         = var.node_pool_02["kubernetes_labels"]
-   disk_size_gb                   = var.node_pool_02["disk_size_gb"]
-   disk_type                      = var.node_pool_02["disk_type"]
-   service_account                = var.service_account
+module "gke_node_pool_02" {
+  source                         = "../modules/google_kubernetes_engine/gke_node_pool/"
+  project_id                     = var.project
+  gke_cluster_name               = module.gke_private_cluster.name
+  node_pool_name                 = "windows-node-pool"
+  region                         = var.region
+  regional                       = var.regional
+  zones                          = module.gke_private_cluster.zones
+  gke_cluster_min_master_version = var.gke_cluster_min_master_version
+  image_type                     = "WINDOWS_LTSC"
+  machine_type                   = var.node_pool_02["machine_type"]
+  preemptible                    = false
+  auto_upgrade                   = false
+  auto_repair                    = true
+  max_pods_per_node              = "100"
+  node_count                     = "1"
+  local_ssd_count                = "0"
+  enable_autoscaling             = true
+  min_node_count                 = var.node_pool_02["min_node_count"]
+  max_node_count                 = var.node_pool_02["max_node_count"]
+  labels                         = var.node_pool_02["kubernetes_labels"]
+  disk_size_gb                   = var.node_pool_02["disk_size_gb"]
+  disk_type                      = var.node_pool_02["disk_type"]
+  service_account                = var.service_account
 }
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
examples/windows_nodepool_with_private_cluster/gke/variables.tf
--- old/examples/windows_nodepool_with_private_cluster/gke/variables.tf
+++ new/examples/windows_nodepool_with_private_cluster/gke/variables.tf
@@ -5,7 +5,7 @@
 variable "network" {}
 
 variable "cluster_name" {
-   default= ""
+  default = ""
 }
 
 variable "zones" {
@@ -21,10 +21,10 @@
   default = "1.16.8-gke.15"
 }
 variable "master_ipv4_cidr_block" {
-  default= ""
+  default = ""
 }
-variable "regional"{
-  default=false
+variable "regional" {
+  default = false
 }
 
 # variable "image_type" {
@@ -81,7 +81,7 @@
 # }
 
 variable "gke_cluster_master_version" {
-  default= ""
+  default = ""
 }
 
 variable "master_authorized_networks" {
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

File: examples/windows_nodepool_with_private_cluster/gke/backend.tf [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/**
```
Comment (bharathkkb) [inline examples/windows_nodepool_with_private_cluster/gke/backend.tf:1]: nit: backends are not needed for examples.

File: examples/windows_nodepool_with_private_cluster/gke/main.tf [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module "gke_private_cluster" {
+  source                     = "../modules/google_kubernetes_engine/gke_private_cluster/"
+  project_id                 = var.project
+  name                       = "test-gke-cluster"
```
Comment (bharathkkb) [inline examples/windows_nodepool_with_private_cluster/gke/main.tf:21]: nit: we should give it a descriptive name, perhaps similar to our existing example naming conventions
https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/26ea28db52ce5797da043a1c56dfb4575f49dac8/examples/simple_regional_private/main.tf#L35

File: examples/windows_nodepool_with_private_cluster/gke/main.tf [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module "gke_private_cluster" {
+  source                     = "../modules/google_kubernetes_engine/gke_private_cluster/"
```
Comment (bharathkkb) [inline examples/windows_nodepool_with_private_cluster/gke/main.tf:19]: nit: is this a custom module?

File: examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_node_pool/main.tf [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This file was automatically generated from a template in ./autogen/main
+
+locals {
```
Comment (bharathkkb) [inline examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_node_pool/main.tf:19]: These should not be added. We should be using one our existing `private-cluster` modules.

File: examples/windows_nodepool_with_private_cluster/gke/main.tf [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  create_service_account     = false
+  service_account            = var.service_account
+  kubernetes_version         = var.gke_cluster_master_version
+  master_authorized_networks = var.master_authorized_networks
+  cluster_resource_labels    = var.gce_labels
+}
+
+#Linux Nodepool 
+module "gke_node_pool_01" {
+  source                         = "../modules/google_kubernetes_engine/gke_node_pool/"
```
Comment (bharathkkb) [inline examples/windows_nodepool_with_private_cluster/gke/main.tf:45]: We should be passing these in via `node_pools= [...]`. Example https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/632#issuecomment-675912953
We can also hardcode as much as possible, like in the linked example.

**Review by bharathkkb — CHANGES_REQUESTED** [review]
Thanks for the PR!
Some initial feedback

File: examples/windows_nodepool_with_private_cluster/gke/main.tf (hunk 3 above)
Comment (bharathkkb) [reply examples/windows_nodepool_with_private_cluster/gke/main.tf:19]: We should not need anything in `./examples/windows_nodepool_with_private_cluster/*`. We can use for instance `"../../modules/beta-private-cluster/"`.

File: examples/windows_nodepool_with_private_cluster/gke/backend.tf (hunk 1 above)
Comment (HarshalRane23) [reply examples/windows_nodepool_with_private_cluster/gke/backend.tf:1]: Done removed backend.tf file

File: examples/windows_nodepool_with_private_cluster/gke/main.tf (hunk 2 above)
Comment (HarshalRane23) [reply examples/windows_nodepool_with_private_cluster/gke/main.tf:21]: done added proper naming convention

File: examples/windows_nodepool_with_private_cluster/gke/main.tf (hunk 3 above)
Comment (HarshalRane23) [reply examples/windows_nodepool_with_private_cluster/gke/main.tf:23]: Yes @bharathkkb  its custom model has taken reference from this same repo, just modified settings to deploy windows nodepool on GKE cluster

File: examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_node_pool/main.tf (hunk 4 above)
Comment (HarshalRane23) [reply examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_node_pool/main.tf:19]: This is custom module so support windows nodepool on GKE cluster, hance added modules section here in repo only.
let me know if that works

Comment (bharathkkb) [reply examples/windows_nodepool_with_private_cluster/modules/google_kubernetes_engine/gke_node_pool/main.tf:19]: Hi @HarshalRane23
We should leverage the existing module and use it in an example with windows nodepools.  An example config can be
```hcl
module "gke" {
  source                            = "../../modules/beta-public-cluster/"
  project_id                        = var.project_id
  name                              = "widows-pool-${var.cluster_name_suffix}"
region                    = var.region
  network                 = var.network
  subnetwork              = var.subnetwork
  ip_range_pods           = var.ip_range_pods
  ip_range_services       = var.ip_range_services
  service_account         = var.compute_engine_service_account
  enable_private_endpoint = true
  enable_private_nodes    = true
  network_policy = false
  master_ipv4_cidr_block  = "172.16.0.0/28"
  enable_shielded_nodes = false
  node_metadata = "EXPOSE"
  identity_namespace = null
  

  master_authorized_networks = [
    {
      cidr_block   = data.google_compute_subnetwork.subnetwork.ip_cidr_range
      display_name = "VPC"
    },
  ]
    node_pools = [
    {
      name               = "linux-node-pool"
      min_count          = 1
      max_count          = 10
      disk_size_gb       = 100
      disk_type          = "pd-standard"
      image_type         = "COS"
      initial_node_count = 1
    },
    {
      name               = "windows-node-pool"
      min_count          = 1
      max_count          = 10
      disk_size_gb       = 100
      disk_type          = "pd-standard"
      image_type         = "WINDOWS_SAC"
      initial_node_count = 1
      enable_integrity_monitoring = false
    }
  ]
}
```

**Pull request closed without merging**